### ISSUE TITLE
Multithreaded version

### ISSUE THREAD

**Issue by attardi** [attardi/wikiextractor#4]
I have released a multithread version of the extractor, that dispatches pages to extract to several worker threads.
However the speed does not seem to improve, but it rather runs more slowly.
Can someone help figure out why?

**Comment by Arkanosis** [comment]
Unfortunately, the reference implementation of Python (ie. CPython) has a GIL, which is a global lock which prevents several threads to perform bytecode interpretation at the same time.

In other words, if you use CPython (and almost everybody does), and the `threading` library (which is the most straightforward, given it's backed by pthreads on Linux), you won't get any improvement over CPU-bound workloads (actually, you'll suffer the overhead of threading and locking, without any parallelization, hence the performance loss). IO-bound workloads, on the other hand can get some improvement because waiting on IO in a thread results in another thread being run.

In the current implementation, I guess the results might actually be better on good CPUs with (very) bad HDDs and a low number of threads, or on some implementation of Python that doesn't have a GIL (Jython, for example) with many threads. There could be some benefit of running the BZip2 compression in another thread (as I believe that it runs outside of the lock, since it's written in C), but probably not enough to counterbalance the overhead.

There's a very nice explanation of the GIL pitfalls here: http://www.dabeaz.com/python/UnderstandingGIL.pdf . Jython get rids of it by relying on the JVM.

**Comment by dirkgr** [comment]
I made a version that forks a new process for every article. I mainly did this because some articles run out of memory, and others go into an infinite loop, and this way I can isolate them. Doing this gives me parallel processing for free.

The way I did it is a complete hack and shouldn't be merged back. That's why I didn't submit a pull request. But the method is still sound. You could either do what I did directly and fork the process, or you could use python's `multiprocessing` module, which will take away a lot of the heavy lifting.

I find that with 32 cores, the bottleneck is how fast it can read the dump, so there is a limit to how much speedup we can get this way.

Oh, and to be sure, one process per article is not a good idea. I just did that because it was easy and I wanted to move on. If there was no problem with runaway processing for an article, every process should handle 100 or 1000 articles. It would be much more efficient that way.

**Comment by gojomo** [comment]
I've hacked together a version that uses multiprocessing.Process, rather than threading.Thread, to get around the GIL issue. I also enabled working from (uncompressed) stdin, so that decompression can be handled in an entirely separate process from the one that's fanning pages out to worker Processes, to make that central step less of a bottleneck.

With these changes, a text-extract of the latest (June 2015) enwiki dump on an 8-core machine went from taking ~500 minutes to ~85 minutes. 

I'd need to clean it up a bit to contribute, as I've done it quick-and-dirty and taken shortcuts specific to my needs. (For example: skipping template-prescan/processing; each worker process just writes to its own one file no matter how large it gets, etc.)

Though right now it's probably getting a benefit from splitting the N output-file compression tasks over separate processes, I might look into changes to hand the <doc>s back to a single process for writing to a single file (or stdout). 

(Two other things I noticed: the sense of the '-s' flag seems the opposite of the documentation, and the daemonization of the threads/processes seemed to create a risk of truncated/corrupted output at the very end, if the parent thread exited before all files flushed/closed. Anyone else notice that in the plain-threads approach as well?)
